crates/trie/sparse/src/state.rs

```diff
@@ -377,56 +384,50 @@ where
     ) -> SparseStateTrieResult<()> {
         let revealed_nodes = self.revealed_storage_paths.entry(account).or_default();
 
-        let FilteredProofNodes {
+        let FilterMappedProofNodes {
+            root_node,
             nodes,
             new_nodes,
             total_nodes: _total_nodes,
             skipped_nodes: _skipped_nodes,
-        } = filter_revealed_nodes(storage_subtree.subtree, revealed_nodes)?;
+        } = filter_map_revealed_nodes(
+            storage_subtree.subtree,
+            revealed_nodes,
+            &storage_subtree.branch_node_hash_masks,
+            &storage_subtree.branch_node_tree_masks,
+        )?;
         #[cfg(feature = "metrics")]
         {
             self.metrics.increment_total_storage_nodes(_total_nodes as u64);
             self.metrics.increment_skipped_storage_nodes(_skipped_nodes as u64);
         }
-        let mut nodes = nodes.into_iter().peekable();
 
-        if let Some(root_node) = Self::validate_root_node_decoded(&mut nodes)? {
+        if let Some(root_node) = root_node {
+            // Perform sanity check.
```

Comment:
this is because we separate the root node from `nodes`?

Reply:
This was a check that `validate_root_node_decoded` used to do, I'm not sure what the context is which makes it necessary, i just moved it here for backwards compatibility.

Reply:
I see, I think this check makes sense anyways